OpenRA.Mods.Common/Traits/Attack/AttackBase.cs

```diff
@@ -211,12 +208,21 @@ public bool HasAnyValidWeapons(Target t)
 
 			// PERF: Avoid LINQ.
 			foreach (var armament in Armaments)
-				if (!armament.OutOfAmmo && armament.Weapon.IsValidAgainst(t, self.World, self))
+			{
+				// If 'onlyTargetCenterWeapons' is true, we don't care about current ammo state, only Weapon.TargetActorCenter
+				var checkIsValidAgainst = onlyTargetCenterWeapons ? armament.Weapon.TargetActorCenter : !armament.OutOfAmmo;
+				if (checkIsValidAgainst && armament.Weapon.IsValidAgainst(t, self.World, self))
 					return true;
+			}
 
 			return false;
 		}
 
+		public virtual WPos GetTargetPosition(WPos pos, Target target, bool onlyTargetCenterWeapons = false)
+		{
+			return HasAnyValidWeapons(target, onlyTargetCenterWeapons) ? target.CenterPosition : target.Positions.PositionClosestTo(pos);
```

Comment:
This looks unclear/wrong to me. It assumes that caller always wants CenterPosition for valid weapon. It is currently called 4 times always with onlyTargetCenterWeapons = true. It doesn't care about armament.Weapon.TargetActorCenter (when called). If you call it with onlyTargetCenterWeapons = false, it will still select CenterPosition.